blockprod/src/detail/mod.rs

```diff
@@ -471,16 +488,10 @@ impl BlockProduction {
                 ));
             }
 
-            // We conservatively use the minimum timestamp here in order to figure out
-            // which transactions are valid for the block.
-            // TODO: Alternatively, we can construct the transaction sequence from
-            // scratch every time a different timestamp is attempted. That is more costly
-            // in terms of computational resources but will allow the node to include more
-            // transactions since the passing time may release some time locks.
             let accumulator = self
                 .collect_transactions(
                     current_tip_index.block_id(),
-                    min_constructed_block_timestamp,
+                    median_time_past,
```

Comment:
I've thought about this a bit more and now I'm not sure that this is the correct fix.
I mean, technically it works, but since there is only one correct value for this parameter, given a specific tip, and since mempool already knows about the current tip, perhaps it shouldn't allow to specify this timestamp at all and calculate it instead?
@iljakuklic I'm wondering what you think about this.

Reply:
It knows the current tip but it does not necessarily know the information about the block it's constructing the transaction sequence for. Passing the information for the candidate block being constructed allows it to include more transactions since some time locked transactions that may have not been valid for inclusion into the current tip may be valid for inclusion to the next block.

Reply:
My only issue with this is that the assumptions are not clear. The value `median_time_past` is not a constant for the whole chain, but depends on the block. As long as it's clear what this value is and what it's based on, this is OK if it works.

Reply:
> Passing the information for the candidate block being constructed allows it to include more transactions since some time locked transactions that may have not been valid for inclusion into the current tip may be valid for inclusion to the next block.

But we can't really pass anything other than this "median time past" here and this value only depends on the parent of the candidate block.

Also, as I can see, mempool does assume that the next block will be on top of the current tip. And it uses this assumption when resolving height-based time locks. So, since timestamp-based locks also only depend on the new block's parent, it looks like they should be treated similarly.